File: packages/core-ts/src/schema.ts
/**
 * Contract validation.
 *
 * The JSON Schemas are the cross-language boundary; this module is the TypeScript side of
 * it. Schema length checks count UTF-16 code units, so a byte guard is applied on top for
 * the fields the contract states in bytes.
 *
 * Version handling is deliberately strict in both directions:
 *
 * - a request declaring a revision this core does not support is `UNSUPPORTED_VERSION`;
 * - a request declaring 1.0 while carrying a 1.1 field, or using a 1.1 operation, is
 *   `INVALID_REQUEST` - an unknown mandatory field is refused, never ignored;
 * - `propose_patch` stays reserved and refused.
 */
import { readFileSync } from "node:fs";
import { join } from "node:path";

import Ajv2020, { type ValidateFunction } from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

import { ShuntError } from "./errors.js";
import {
  DEFAULT_LIMITS,
  V11_ONLY_OPERATIONS,
  V11_ONLY_REQUEST_FIELDS,
  contractsDir,
  supportedRequestVersion,
} from "./limits.js";
import { utf8Length } from "./textindex.js";

const ajv = addFormats(new Ajv2020({ allErrors: false, strict: false }));
const cache = new Map<string, ValidateFunction>();

export const READ_OPERATIONS: ReadonlySet<string> = new Set(["read"]);
export const INSPECT_OPERATIONS: ReadonlySet<string> = new Set(["inspect"]);
export const STATS_OPERATIONS: ReadonlySet<string> = new Set(["stats"]);
export const ALL_OPERATIONS: ReadonlySet<string> = new Set(["read", "inspect", "stats"]);

function validator(name: string): ValidateFunction {
  let fn = cache.get(name);
  if (!fn) {
    const schema = JSON.parse(readFileSync(join(contractsDir(), name), "utf8")) as object;
    fn = ajv.compile(schema);
    cache.set(name, fn);
  }
  return fn;
}

export function requestValidator(): ValidateFunction {
  return validator("request.schema.json");
}

export function envelopeValidator(): ValidateFunction {
  return validator("envelope.schema.json");
}

export function toolArgsValidator(): ValidateFunction {
  return validator("tool-args.schema.json");
}

export interface ReaderRequest {
  schema_version: string;
  request_id: string;
  operation: "read";
  question: string;
  sources: Array<{
    source_id: string;
    snapshot_id: string;
    selector: Record<string, unknown>;
  }>;
  budgets: { max_chunks: number; max_answer_bytes: number; deadline_ms: number };
  refined?: boolean;
}

export interface InspectRequest {
  schema_version: string;
  request_id: string;
  operation: "inspect";
  source_id: string;
  snapshot_id: string;
  selector: Record<string, unknown>;
  budgets: { max_result_bytes: number; max_scan_lines: number };
  cursor?: string;
}

export interface StatsRequest {
  schema_version: string;
  request_id: string;
  operation: "stats";
  page?: number;
  page_size?: number;
}

export type CoreRequest = ReaderRequest | InspectRequest | StatsRequest;

/**
 * Validate one core request. `operations` is the set this call site accepts, so the reader
 * cannot be handed a stats request and the stats path cannot be handed a question.
 */
export function validateRequest(
  request: unknown,
  operations: ReadonlySet<string> = READ_OPERATIONS,
): CoreRequest {
  if (typeof request !== "object" || request === null) {
    throw new ShuntError("INVALID_REQUEST", "NOT_OBJECT", false);
  }
  const req = request as Record<string, unknown>;
  const version = req["schema_version"];
  if (!supportedRequestVersion(version)) {
    throw new ShuntError("UNSUPPORTED_VERSION", "BAD_SCHEMA_VERSION", false);
  }
  const operation = req["operation"];
  if (operation === "propose_patch") {
    // Reserved for a future writer contract. v1 refuses it as an unsupported operation
    // rather than treating it as an unknown enum value.
    throw new ShuntError("INVALID_REQUEST", "WRITER_OPERATION_UNSUPPORTED", false);
  }
  if (typeof operation !== "string" || !ALL_OPERATIONS.has(operation)) {
    throw new ShuntError("INVALID_REQUEST", "UNKNOWN_OPERATION", false);
  }
  if (!operations.has(operation)) {
    throw new ShuntError("INVALID_REQUEST", "OPERATION_NOT_ACCEPTED_HERE", false);
  }
  if (version === "1.0") {
    if (V11_ONLY_OPERATIONS.has(operation)) {
      throw new ShuntError("INVALID_REQUEST", "OPERATION_REQUIRES_1_1", false);
    }
    for (const key of Object.keys(req)) {
      if (V11_ONLY_REQUEST_FIELDS.has(key)) {
        throw new ShuntError("INVALID_REQUEST", "FIELD_REQUIRES_1_1", false);
      }
    }
  }
  if (!requestValidator()(request)) {
    throw new ShuntError("INVALID_REQUEST", "SCHEMA_VIOLATION", false);
  }
  if (operation === "read") assertQuestion(String(req["question"] ?? ""));
  return request as unknown as CoreRequest;
}

/** Validate the arguments an agent passed to one of the three escape-hatch tools. */
export function validateToolArgs(args: unknown): Record<string, unknown> {
  if (typeof args !== "object" || args === null) {
    throw new ShuntError("INVALID_REQUEST", "NOT_OBJECT", false);
  }
  if (!toolArgsValidator()(args)) {
    throw new ShuntError("INVALID_REQUEST", "TOOL_ARGS_VIOLATION", false);
  }
  const question = (args as Record<string, unknown>)["question"];
  if (typeof question === "string") assertQuestion(question);
  return args as Record<string, unknown>;
}

function assertQuestion(question: string): void {
  if (utf8Length(question) > DEFAULT_LIMITS.maxQuestionBytes) {
    throw new ShuntError("INVALID_REQUEST", "QUESTION_OVER_BYTE_CAP", false);
  }
  if (question.trim().length === 0) {
    throw new ShuntError("INVALID_REQUEST", "EMPTY_QUESTION", false);
  }
}

export function validateEnvelope(envelope: unknown): boolean {
  return envelopeValidator()(envelope) === true;
}

export function envelopeErrors(envelope: unknown): string[] {
  const validate = envelopeValidator();
  validate(envelope);
  return (validate.errors ?? []).map((error) => error.message ?? "invalid");
}

File: packages/core-py/src/context_shunt/schema.py
"""Contract validation.

The JSON Schemas are the cross-language boundary; this module is the Python side of it.
Schema length checks count UTF-16 code units, so a byte guard is applied on top for the
fields where the contract is stated in bytes.

Version handling is deliberately strict in both directions:

* a request declaring a revision this core does not support is ``UNSUPPORTED_VERSION``;
* a request declaring 1.0 while carrying a 1.1 field, or using a 1.1 operation, is
  ``INVALID_REQUEST`` - an unknown mandatory field is refused, never ignored;
* ``propose_patch`` stays reserved and refused.
"""

from __future__ import annotations

import json
from functools import cache
from typing import Any

from jsonschema import Draft202012Validator

from .errors import ShuntError
from .limits import (
    CONTRACTS_DIR,
    DEFAULT_LIMITS,
    V11_ONLY_OPERATIONS,
    V11_ONLY_REQUEST_FIELDS,
    supported_request_version,
)

READ_OPERATIONS = frozenset({"read"})
ALL_OPERATIONS = frozenset({"read", "inspect", "stats"})


@cache
def _validator(name: str) -> Draft202012Validator:
    with (CONTRACTS_DIR / name).open("rb") as fh:
        return Draft202012Validator(json.load(fh))


def request_validator() -> Draft202012Validator:
    return _validator("request.schema.json")


def envelope_validator() -> Draft202012Validator:
    return _validator("envelope.schema.json")


def tool_args_validator() -> Draft202012Validator:
    return _validator("tool-args.schema.json")


def validate_request(
    request: Any, *, operations: frozenset[str] = READ_OPERATIONS
) -> dict[str, Any]:
    """Validate one core request.

    ``operations`` is the set this call site accepts, so the reader cannot be handed a
    stats request and the stats path cannot be handed a question.
    """
    if not isinstance(request, dict):
        raise ShuntError("INVALID_REQUEST", "NOT_OBJECT", retryable=False)
    version = request.get("schema_version")
    if not supported_request_version(version):
        raise ShuntError("UNSUPPORTED_VERSION", "BAD_SCHEMA_VERSION", retryable=False)
    operation = request.get("operation")
    if operation == "propose_patch":
        # Reserved for a future writer contract. v1 refuses it as an unsupported
        # operation rather than treating it as an unknown enum value.
        raise ShuntError("INVALID_REQUEST", "WRITER_OPERATION_UNSUPPORTED", retryable=False)
    if operation not in ALL_OPERATIONS:
        raise ShuntError("INVALID_REQUEST", "UNKNOWN_OPERATION", retryable=False)
    if operation not in operations:
        raise ShuntError("INVALID_REQUEST", "OPERATION_NOT_ACCEPTED_HERE", retryable=False)
    if version == "1.0":
        if operation in V11_ONLY_OPERATIONS:
            raise ShuntError("INVALID_REQUEST", "OPERATION_REQUIRES_1_1", retryable=False)
        if set(request) & V11_ONLY_REQUEST_FIELDS:
            raise ShuntError("INVALID_REQUEST", "FIELD_REQUIRES_1_1", retryable=False)
    if request_validator().is_valid(request):
        _byte_guards(request)
        return request
    raise ShuntError("INVALID_REQUEST", "SCHEMA_VIOLATION", retryable=False)


def validate_tool_args(args: Any) -> dict[str, Any]:
    """Validate the arguments an agent passed to one of the three escape-hatch tools."""
    if not isinstance(args, dict):
        raise ShuntError("INVALID_REQUEST", "NOT_OBJECT", retryable=False)
    if not tool_args_validator().is_valid(args):
        raise ShuntError("INVALID_REQUEST", "TOOL_ARGS_VIOLATION", retryable=False)
    question = args.get("question")
    if isinstance(question, str):
        _assert_question(question)
    return args


def _byte_guards(request: dict[str, Any]) -> None:
    if request.get("operation") != "read":
        return
    _assert_question(request.get("question", ""))


def _assert_question(question: str) -> None:
    if len(question.encode("utf-8")) > DEFAULT_LIMITS.max_question_bytes:
        raise ShuntError("INVALID_REQUEST", "QUESTION_OVER_BYTE_CAP", retryable=False)
    if not question.strip():
        raise ShuntError("INVALID_REQUEST", "EMPTY_QUESTION", retryable=False)


def validate_envelope(envelope: Any) -> bool:
    return envelope_validator().is_valid(envelope)


def envelope_errors(envelope: Any) -> list[str]:
    return [e.message for e in envelope_validator().iter_errors(envelope)]
